lib/src/sql/regex.rs

```diff
@@ -4,59 +4,120 @@ use nom::bytes::complete::escaped;
 use nom::bytes::complete::is_not;
 use nom::character::complete::anychar;
 use nom::character::complete::char;
-use serde::{Deserialize, Serialize};
-use std::fmt;
-use std::ops::Deref;
+use serde::{
+	de::{self, Visitor},
+	Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::cmp::Ordering;
+use std::fmt::Debug;
+use std::fmt::{self, Display, Formatter};
+use std::hash::{Hash, Hasher};
 use std::str;
+use std::str::FromStr;
 
 pub(crate) const TOKEN: &str = "$surrealdb::private::sql::Regex";
 
-#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd, Deserialize, Hash)]
-pub struct Regex(pub(super) String);
+#[derive(Clone)]
+pub struct Regex(pub(super) regex::Regex);
 
-impl From<&str> for Regex {
-	fn from(r: &str) -> Self {
-		Self(r.replace("\\/", "/"))
+impl Regex {
+	// Deref would expose `regex::Regex::as_str` which wouldn't have the '/' delimiters.
+	pub fn regex(&self) -> &regex::Regex {
+		&self.0
 	}
 }
 
-impl Deref for Regex {
-	type Target = String;
-	fn deref(&self) -> &Self::Target {
-		&self.0
+impl FromStr for Regex {
+	type Err = <regex::Regex as FromStr>::Err;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		regex::Regex::new(&s.replace("\\/", "/")).map(Self)
 	}
 }
 
-impl fmt::Display for Regex {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(f, "/{}/", &self.0)
+impl PartialEq for Regex {
+	fn eq(&self, other: &Self) -> bool {
+		self.0.as_str().eq(other.0.as_str())
 	}
 }
 
-impl Regex {
-	pub fn regex(&self) -> Option<regex::Regex> {
-		regex::Regex::new(&self.0).ok()
+impl Eq for Regex {}
+
+impl Ord for Regex {
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.0.as_str().cmp(other.0.as_str())
+	}
+}
+
+impl PartialOrd for Regex {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Hash for Regex {
+	fn hash<H: Hasher>(&self, state: &mut H) {
+		self.0.as_str().hash(state);
+	}
+}
+
+impl Debug for Regex {
+	fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+		Display::fmt(self, f)
+	}
+}
+
+impl Display for Regex {
+	fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+		write!(f, "/{}/", &self.0)
 	}
 }
 
 impl Serialize for Regex {
 	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
 	where
-		S: serde::Serializer,
+		S: Serializer,
 	{
 		if is_internal_serialization() {
-			serializer.serialize_newtype_struct(TOKEN, &self.0)
+			serializer.serialize_newtype_struct(TOKEN, self.0.as_str())
 		} else {
 			serializer.serialize_none()
 		}
 	}
 }
 
+pub struct RegexVisitor;
```

Comment:
I know this module is private so I'm probably just nitpicking here but this doesn't need to be public, right? I'm worried that it might accidentally leak in future. In fact, the convention in Serde is to put `Visitor` implementations inside the `Deserialize` implementation itself but I think leaving it outside like this is OK too.